omp/reorder/rcm_kernels.cpp

```diff
@@ -0,0 +1,908 @@
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include "core/reorder/rcm_kernels.hpp"
+#include "core/base/allocator.hpp"
+#include "core/components/prefix_sum.hpp"
+
+
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <queue>
+#include <utility>
+#include <vector>
+
+
+#include <omp.h>
+
+
+#include <ginkgo/config.hpp>
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/math.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/matrix/coo.hpp>
+#include <ginkgo/core/matrix/csr.hpp>
+#include <ginkgo/core/matrix/permutation.hpp>
+#include <ginkgo/core/matrix/sparsity_csr.hpp>
+
+
+namespace gko {
+namespace kernels {
+namespace omp {
+/**
+ * @brief The reordering namespace.
+ *
+ * @ingroup reorder
+ */
+namespace rcm {
+
+
+#ifdef __x86_64__
+#if defined(__GNUG__) || defined(__clang__)
+#include <immintrin.h>
+#endif
+#define GKO_MM_PAUSE() _mm_pause()
+#else
+#define GKO_MM_PAUSE()  // No equivalent instruction.
+#endif
+
+template <typename IndexType>
+void get_degree_of_nodes(std::shared_ptr<const OmpExecutor> exec,
+                         const size_type num_vertices,
+                         const IndexType *const row_ptrs,
+                         IndexType *const degrees)
+{
+#pragma omp parallel for
+    for (auto i = 0; i < num_vertices; ++i) {
+        degrees[i] = row_ptrs[i + 1] - row_ptrs[i];
+    }
+}
+
+GKO_INSTANTIATE_FOR_EACH_INDEX_TYPE(GKO_DECLARE_RCM_GET_DEGREE_OF_NODES_KERNEL);
+
+
+// This constant controls how many nodes can be dequeued from the
+// UbfsLinearQueue at once at most. Increasing it reduces lock contention and
+// "unneccesary work", but disturbs queue ordering, generating extra work.
+constexpr auto chunk_bound = 512;
+
+
+template <typename IndexType>
+struct UbfsLinearQueue {
+    vector<IndexType> arr;
+    IndexType head;
+    IndexType tail;
+
+    // Those two are only there for safety, they are barely ever contended,
+    // due to work chunking.
+    omp_lock_t read_lock;
+    omp_lock_t write_lock;
+
+
+    UbfsLinearQueue(std::shared_ptr<const OmpExecutor> exec, size_type capacity)
+        : arr(vector<IndexType>(capacity, exec)), head(0), tail(0)
+    {
+        omp_init_lock(&read_lock);
+        omp_init_lock(&write_lock);
+    }
+
+    ~UbfsLinearQueue()
+    {
+        omp_destroy_lock(&read_lock);
+        omp_destroy_lock(&write_lock);
+    }
+
+    UbfsLinearQueue(UbfsLinearQueue &other) = delete;
+    UbfsLinearQueue &operator=(const UbfsLinearQueue &other) = delete;
+
+    /**
+     * Copies a chunk of nodes back into the work queue,
+     * in a thread-safe manner.
+     */
+    void enqueue_chunk(const IndexType *const chunk, size_type n)
+    {
+        const auto data = &arr[0];
+
+        omp_set_lock(&write_lock);
```

Comment:
Well, the reason is that this program is - or at least tries to be - OpenMP standard compliant. The standard masterfully ignores the existence of `std::` concurrency, all interactions are unspecified. I did not want to rely on the compilers doing the right thing here.

Reply:
As far as I understand it, any synchronization primitives used in the C++ standard library are supposed to work independent of the underlying threading model (OpenMP, pthread, std::thread, TBB, ...). While OpenMP doesn't make any guarantees on the interactions from the OpenMP standard side, a mutex lock and unlock is nothing else but a function call to the compiler (probably pthread_mutex_*lock) under the hood, so something would seriously have to go wrong for this to cause any issues, especially since these functions need to be inherently thread-safe.

Reply:
Well, there we have our interaction: pthreads, and whatever library OpenMP relies on, usually also pthreads. This might very well work on x86 for common scenarios, it is entirely unspecified though. To cite [this answer](https://stackoverflow.com/a/13839719/8511998), which is still valid to this day: "supporting such functionality is up to the implementers". I would be especially cautious with compiling for Power9 and ARM.
Perhaps a guard wrapper around those locks would be a better solution?

Reply:
To raise another point, even if `std::mutex` doesn't belong to the rather nebulous "concurrency" category, it definitely belongs in the "atomics and memory model" category.

Reply:
I also think it is supposed to work since simple locks should work without knowing which type of threading model is used.
However, I understand your point that it is officially not supported, and therefore in the undefined behavior category.

I would suggest to put OpenMP locks into a mutex class in `omp/components` to enable RAII, which can simply look like that:
```c++
struct omp_mutex {
    omp_mutex() { omp_init_lock(&lock_); }
    ~omp_mutex() { omp_destroy_lock(&lock_); }

    omp_mutex(const omp_mutex &) = delete;
    omp_mutex(omp_mutex &&) = delete;
    omp_mutex &operator=(const omp_mutex &) = delete;
    omp_mutex &operator=(omp_mutex &&) = delete;

    void lock() { omp_set_lock(&lock_); }

    void unlock() { omp_unset_lock(&lock_); }

private:
    omp_lock_t lock_;
};
```
Aquiring a lock is now possible with:
```c++
omp_mutex mutex{};
...
std::lock_guard<omp_mutex> guard{mutex};
```
`std::lock_guard` simply calls `mutex.lock()` when created, and `mutex.unlock()` when destroyed. That way, you are only using OpenMP locks, but with all the benefits of RAII.